### PR TITLE
listener component refactor

### PR DESCRIPTION
**Pull request by mattklein123**
Add another factory that is used to create the list of filter chain
factory callbacks. This will make LDS tests simpler to write.

### CODE REVIEW

**Comment by mattklein123** [comment]
@lyft/network-team note this currently has https://github.com/lyft/envoy/pull/1172 in it because of the way I'm stacking branches. Will rebase once that has merged.

**Comment by mattklein123** [comment]
Nevermind I just rebased directly on master so this is fine to review as is.

File: source/server/listener_manager_impl.cc [hunk 1]
```diff
@@ -102,8 +66,53 @@ ListenerImpl::ListenerImpl(Instance& server, ListenSocketFactory& factory, const
...
+}
+
+Network::ListenSocketPtr
+ProdListenerComponentFactory::createListenSocket(Network::Address::InstanceConstSharedPtr address,
+                                                 bool bind_to_port) {
+  // For each listener config we share a single TcpListenSocket among all threaded listeners.
+  // UdsListenerSockets are not managed and do not participate in hot restart as they are only
+  // used for testing. First we try to get the socket from our parent if applicable.
+  ASSERT(address->type() == Network::Address::Type::Ip);
+  std::string addr = fmt::format("tcp://{}", address->asString());
```
Comment (RomanDzhabarov) [inline source/server/listener_manager_impl.cc]: const std::string ?

File: include/envoy/server/listener_manager.h [hunk 2]
```diff
@@ -3,26 +3,37 @@
...
+  virtual Network::ListenSocketPtr
+  createListenSocket(Network::Address::InstanceConstSharedPtr address, bool bind_to_port) PURE;
+
+  /**
+   * Creates a list of filter factories.
+   * @param filters supplies the JSON configuration.
+   * @param context supplies the factory creation context.
+   * @return std::list<Configuration::NetworkFilterFactoryCb> the list of filter factories.
+   */
+  virtual std::list<Configuration::NetworkFilterFactoryCb>
```
Comment (RomanDzhabarov) [inline include/envoy/server/listener_manager.h]: any reason it's not std::vector?

**Comment by mattklein123** [comment]
@RomanDzhabarov updated

**Pull request merged**